### ISSUE TITLE
Should tss_limit just be sizeof(gTss)?

### ISSUE THREAD

**Issue by PiJoules** [ozkl/soso#7]
tl;dr I think the `tss_limit` should refer to the size of TSS rather than the end address of the TSS.
____

I've been using this project as a reference for my own toy OS and recently ran into an issue where I was somehow able to run I/O instructions (`in`/`out`) in userspace. I was perplexed why this was happening since I was able to assert that I'm running in protected mode, and the current privilege level (3) is greater than the I/O privilege level (0).

I dug deeper into this and found out that, even in this situation, an exception won't be thrown if the I/O permission bit corresponding to a given I/O port is marked as accessible in the I/O permission bitmap [1]. This bitmap seems to be only relevant though if we want certain ports to be accessible regardless of CPL. In my case, I didn't want to use this bitmap and only depend on the CPL and IOPL.

This could be disabled though if the I/O map base in the TSS was set to a value greater than or equal to the TSS limit, which is set in the TSS descriptor for the GDT. Initially, I left this value as zero, but setting it to `sizeof(gTss)` didn't seem to fix my problem. What I did notice though was that the limit was set as the base of the TSS + the size of the TSS. Lo and behold, setting this limit to `sizeof(gTss)` fixed my issue and I could reliably fault in userspace.

It seems that there are also other examples on OSDev where the this limit used in the descriptor to the size of the TSS rather than the end address. [2][3] Sharing this because it took me some time to figure this out and wanted to save folks some time debugging if this is an underlying issue :)

[1] https://c9x.me/x86/html/file_module_x86_id_222.html
[2] https://wiki.osdev.org/GDT_Tutorial#Flat_Setup
[3] The `write_tss` function in https://wiki.osdev.org/Getting_to_Ring_3

**Comment by ozkl** [comment]
I was testing my userspace with privileged instructions like CLI and when getting "General Protection Fault" I was thinking it was OK.

I've just learnt this x86 feature: granting I/O permissions to userspace.

When I first started this project, I followed JamesM [1] tutorial which still points that the limit is end address of the TSS.

Thanks for sharing this :)
I will definitely check this out!

[1] http://www.jamesmolloy.co.uk/tutorial_html/10.-User%20Mode.html
